### DISCUSSION TITLE
Wrong Type Signature In Documentation Of `createSignal`

### DISCUSSION THREAD

**Discussion by fweth** [solidjs/solid#864] [General]
In the documentation, the type signature of `createSignal` expects the setter of type `(v: T) => T` but this does not reflect the case where the setter is a function, described later as

>As its only argument, the setter takes either the new value for the signal, or a function that maps the last value of the signal to a new value. The setter also returns the updated value.

In the source code, this case seems to be included:

    export type Setter<T> = undefined extends T
      ? <U extends T>(value?: (U extends Function ? never : U) | ((prev?: T) => U)) => U
      : <U extends T>(value: (U extends Function ? never : U) | ((prev: T) => U)) => U;

Props to Reddit user [/u/shaberman](https://www.reddit.com/r/typescript/comments/sy40r3/comment/hxw75bu/) for pointing me to the source.

**Comment by ryansolid** [comment]
Yeah it is out of date. Most of the types in the docs are an approximation. But adding the function support was in like 1.0-rc.7 or something right before final release after I had written all the docs.

**Reply by fweth** [reply]
Yes, the actual type is really a bit complicated (because, as far as I can see, it says that the function clause can't be used when the type of the signal is a function already), I wouldn't know how to port that to the documentation in an elegant way.
